adapters/src/controller/mod.rs

```diff
@@ -624,70 +889,144 @@ mod test {
         SizeOf,
         Encode,
         Decode,
+        Arbitrary,
     )]
     struct TestStruct {
+        id: u32,
         b: bool,
-        f: F32,
-        s: Option<String>,
+        i: Option<i64>,
+        s: String,
     }
 
-    /*impl TestStruct {
-        fn new(b: bool, f: F32, s: Option<String>) -> Self {
-            Self { b, f, s }
-        }
-    }*/
+    fn test_data(size: usize) -> impl Strategy<Value = Vec<TestStruct>> {
+        collection::vec(any::<TestStruct>(), 0..=size)
+    }
 
     fn test_circuit(workers: usize) -> (DBSPHandle, Catalog) {
-        let (circuit, (input, _output)) = Runtime::init_circuit(workers, |circuit| {
+        let (circuit, (input, output)) = Runtime::init_circuit(workers, |circuit| {
             let (input, hinput) = circuit.add_input_zset::<TestStruct, i32>();
 
-            let houtput = input.integrate().output();
+            let houtput = input.output();
             (hinput, houtput)
         })
         .unwrap();
 
         let mut catalog = Catalog::new();
-        catalog.register_input_zset("test_input1", input);
+        catalog.register_input_zset_handle("test_input1", input);
+        catalog.register_output_batch_handle("test_output1", output);
 
         (circuit, catalog)
     }
 
-    #[test]
-    fn csv_file_test() {
-        let (circuit, catalog) = test_circuit(4);
-
-        let temp_file = NamedTempFile::new().unwrap();
-
-        let config_str = format!(
-            r#"
+    // TODO: Parameterize this with config string, so we can test different
+    // input/output formats and transports when we support more than one.
+    proptest! {
+        #![proptest_config(ProptestConfig::with_cases(30))]
+        #[test]
+        fn proptest_csv_file(
+            data in test_data(5000),
+            min_batch_size_records in 1..100usize,
```

Comment:
this generates a random configuration within these parameters?
That's an interesting tool.
Hope that you can somehow find out what they were if something goes wrong.

Reply:
yes, it's part of the proptest magic.  It tries to produce a minimal failing example.  And records it in a file that can be checked into the repo.